settings: Adjust label and field-hint layout.

karlstolley opened the pull request:
This draft PR displays non-checkbox labels and field hints, where present, spaced equally from the inputs that follow them.

Thoughts, questions, & to-dos:

- [x] There are labels with a `title` class, but no associated style with them. (E.g., on the Emoji theme label); only the Language field has it styled, and that could probably be cleaned up and restyled with the new, shared label class anyway
- [x] Replace the `dropdown-title` class as I go, but save cleanup of the CSS until all instances are accounted for
- [x] `<label>` inherits a generous 20px line-height for its diminutive 14px font-size, so conceivably we could do a negative top-margin on field hint of `-4.5px` (multiply the value by `-1.5`), though that would likely need to be rethought on any future line-height work

[CZO discussion](https://chat.zulip.org/#narrow/stream/101-design/topic/field.20label.20and.20hint.20positions/near/1758968)

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

_Profile labels and hints_

| Before | After |
| --- | --- |
| ![settings-label-before](https://github.com/zulip/zulip/assets/170719/5fcf1aec-fa07-4f81-a58b-3089e432186d) | ![settings-label-after](https://github.com/zulip/zulip/assets/170719/1e78cb1e-1067-43b7-94f3-7917b59d7ac9) |

_Profile labels, errors, and hints_

| Before | After |
| --- | --- |
| ![settings-label-with-error-before](https://github.com/zulip/zulip/assets/170719/172c801e-e4dd-4edc-bffd-b124adee8284) | ![settings-label-with-error-after](https://github.com/zulip/zulip/assets/170719/5152937c-5ea0-4c9f-967a-ed7f75e8cd1c) |

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sahil839 commented:
Changes look good. The first two suggestions in the PR description also seems good.

I do not get the third one. I guess currently the label hint and inputs are equidistant. So, after that change, will they be equidistant or not?

karlstolley commented:
@sahil839 said:

> Changes look good. The first two suggestions in the PR description also seems good.
> 
> I do not get the third one. I guess currently the label hint and inputs are equidistant. So, after that change, will they be equidistant or not?

The label and the hint would be cinched a little closer together. That would buy back a little bit of the additional vertical space introduced by uniform bottom margins after all labels, and from a Gestalt perspective, help the label and hint function as a single unit--making for a bit less visual clutter for users scanning the page for the setting they're after.

sahil839 commented:
Yeah, having lable and hint a bit more closer sounds better.

karlstolley commented:
Okay @sahil839 -- I've fixed the merge conflicts here and also added a prep commit that tidies up what's now the `.settings-profile-user-field-hint` class. I've also cinched up the space between the label and the hint, though I note that the generous 20px of line-height on `<label>` is a Bootstrap thing. This now looks like this:

![Screen Shot 2024-03-27 at 16 12 36](https://github.com/zulip/zulip/assets/170719/137c98f3-d6c9-4360-a8e1-de37b1804367)


Are you interested in carrying this PR forward and just applying the new classes here where they belong across the settings modal?

I'm thinking the `for`- and `id`-attribute fixes that are needed probably belong as their own PR, if you'd like to pick that up, too.

sahil839 commented:
Some of the things that I am not sure how to handle -
- Do we need the margin below "Information density settings" label? It does not have an input or button directly below it.
- In invite modal, the label for radio buttons for email and multiuse invite already has some space below it and we are anyways going to remove that and instead use tabs, so I guess we can leave that as it is for now.
![image](https://github.com/zulip/zulip/assets/35494118/d50488ef-20cb-4ce9-966b-486c90b55f7a)
- For the custom profile fields, there is an `<hr>` element below the "Field choices" heading for "List of options" type fields in the field edit form (which results in some space between label and inputs) but not in field creation form, so I guess we can make that consistent.

Custom profile field creation
![image](https://github.com/zulip/zulip/assets/35494118/81ea8b52-bf8b-47b3-8be0-50cddd2005e1)

Custom profile field edit form
![image](https://github.com/zulip/zulip/assets/35494118/f937ff38-3e1c-47d9-97bc-244fcdcf5591)

karlstolley commented:
Here are some thoughts, @sahil839:

* The information-density stuff is dev-only and temporary, but even if nothing else changes "Information density settings" should be something other than a `<label>` element
* Radio button plan sounds good to me on the Invite modal
* I'm all for consistency...if the rule stays, can that be accomplished in CSS only with, say, `border-bottom`?

sahil839 commented:
And yes, the bot avatar case in bot creation form is a bit different especially in bot edit form.

In creation form we show a button to add avatar initially and then the selected image, There is a bottom margin of 3px below the "Avatar" label for both the cases.

In bot edit form, we do not show the initial avatar of the bot and there is some space between the "Avatar" label and "Change avatar" button. But if we change the avatar, the new avatar is shown and we can add the bottom margin for that case. So, for now I have added the margin below the label, but that means the space between "Change avatar" and label initially after opening the modal increases.

We should fix this completely by showing the initial avatar as well, but till then we can decide whether we just want to avoid adding the margin below "Avatar" label for now.

sahil839 commented:
> I'm all for consistency...if the rule stays, can that be accomplished in CSS only with, say, border-bottom?

We can do that by adding a bottom border to label and a top border to the `div` element after it to match the current styling of `hr` element.

But I read through some history to find the code which added the `hr` divider and I found that it was added in the initial implementaion of the UI to support "List of options" custom profile fields. So, just wanted to make sure whether we still need a `hr` element below the label. @alya FYI.

alya commented:
That <hr> seems extraneous to me; I think we can drop it.

sahil839 commented:
Added commit to remove the hr element.

Regarding bot avatar, the bug in edit form is that the current avatar is not shown only when it is not a user uploaded avatar. If it is a user uploaded avatar. So, I have just kept the changes to add the margin below "Avatar" label for now.

It results in some space added to already existing extra space between label and "Change avatar" button in edit form if the bot does not have user uploaded avatar, but it already had some extra space before as well so should be fine in short term. Will try to get a fix for it soon.

karlstolley commented:
Code changes look good to me, @sahil839! Can you fix the merge conflicts and I'll give this one more look in dev? @alya might want to do a pass on it, too, with her eagle eyes for subtle differences :)

alya commented:
I dunno, seems fine to me. This isn't really the sort of thing I have eagle eyes for. ;)

timabbott commented:
Took a look through this and it looks great, merged, thanks @karlstolley and @sahil839!
